src/Build/BuildCheck/Infrastructure/CustomConfigurationData.cs

```diff
@@ -55,10 +65,40 @@ public override bool Equals(object? obj)
             return false;
         }
 
-        return Equals((CustomConfigurationData)obj);
-    }
+        var customConfigObj = (CustomConfigurationData) obj;
+
+        if(customConfigObj.RuleId != RuleId)
+        {
+            return false;
+        }
 
-    protected bool Equals(CustomConfigurationData other) => Equals(ConfigurationData, other.ConfigurationData);
+        // validate keys and values
+        if (customConfigObj.ConfigurationData != null && ConfigurationData != null)
+        {
+            if (!customConfigObj.ConfigurationData.Keys.SequenceEqual(ConfigurationData.Keys))
+            {
+                return false;
+            }
+
+            var keys = customConfigObj.ConfigurationData.Keys;
+            foreach (var key in keys)
+            {
+                if (customConfigObj.ConfigurationData[key] != ConfigurationData[key])
+                {
+                    return false;
+                }
+            }
+        }else if (customConfigObj.ConfigurationData == null && ConfigurationData == null)
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+
+        return true;
+    }
 
     public override int GetHashCode() => (ConfigurationData != null ? ConfigurationData.GetHashCode() : 0);
```

Comment:
`GetHashCode` should use all instance data to calculate the code. More importantly, instances that are considered equal per the `Equals` implementation must have the same hash code. Note that `Dictionary<K,V>` doesn't override `GetHashCode` so the value you are returning here is _not_ derived from the contents of the dictionary.

This snippet demonstrates the issue:
https://dotnetfiddle.net/Y8k7Cm

Reply:
Agree, I'll update the implementation for future usage (in case will be needed). 
Since the custom configuration data is not used in the Dictionary in a 'key'  role this is not reflected anyhow in the tests.

Reply:
Implementation updated.